pkg/candevice/device_linux.go

```diff
@@ -0,0 +1,425 @@
+	Clock            struct{ unix.CANClock }
+	CtrlMode         struct{ unix.CANCtrlMode }
+	BusErrorCounters struct{ unix.CANBusErrorCounters }
+	Stats            struct{ unix.CANDeviceStats }
+)
+
+type ifInfoMsg struct{ unix.IfInfomsg }
+
+type Info struct {
+	BitTiming        BitTiming
+	BitTimingConst   BitTimingConst
+	Clock            Clock
+	CtrlMode         CtrlMode
+	BusErrorCounters BusErrorCounters
+
+	State     uint32
+	RestartMs uint32
+}
+
+type linkInfoMsg struct {
+	linkType string
+	info     Info
+	stats    Stats
+}
+
+type Device struct {
+	ifname string
+	index  int32
+	li     linkInfoMsg
+	ifi    ifInfoMsg
+}
+
+// Creates a handle to a CAN device specified by name, e.g. can0.
+func New(deviceName string) (*Device, error) {
+	d := &Device{}
+	iface, err := net.InterfaceByName(deviceName)
+	if err != nil {
+		return nil, err
+	}
+	d.index = int32(iface.Index)
+	if err := d.updateInfo(); err != nil {
+		return nil, err
+	}
+	return d, nil
+}
+
+func (d *Device) IsUp() (bool, error) {
+	if err := d.updateInfo(); err != nil {
+		return false, err
+	}
+	if d.ifi.Flags&unix.IFF_UP != 0 {
+		return true, nil
+	}
+	return false, nil
+}
+
+// corresponds to "ip link set up".
+func (d *Device) SetUp() error {
+	c, err := netlink.Dial(unix.NETLINK_ROUTE, &netlink.Config{})
+	if err != nil {
+		return fmt.Errorf("couldn't dial netlink socket: %w", err)
+	}
+	defer c.Close()
+
+	ifi := &ifInfoMsg{}
+	ifi.Index = d.index
+	ifi.Flags = unix.IFF_UP
+	ifi.Change = unix.IFF_UP
+	req, err := d.newRequest(unix.RTM_NEWLINK, ifi)
+	if err != nil {
+		return fmt.Errorf("couldn't create netlink request: %w", err)
+	}
+
+	res, err := c.Execute(req)
+	if err != nil {
+		return fmt.Errorf("couldn't set link up: %w", err)
+	}
+	if len(res) > 1 {
+		return fmt.Errorf("expected 1 message, got %d", len(res))
+	}
+	return nil
+}
+
+// corresponds to "ip link set down".
+func (d *Device) SetDown() error {
+	c, err := netlink.Dial(unix.NETLINK_ROUTE, &netlink.Config{})
+	if err != nil {
+		return fmt.Errorf("couldn't dial netlink socket: %w", err)
+	}
+	defer c.Close()
+
+	ifi := &ifInfoMsg{}
+	ifi.Index = d.index
+	ifi.Flags = 0
+	ifi.Change = unix.IFF_UP
+	req, err := d.newRequest(unix.RTM_NEWLINK, ifi)
+	if err != nil {
+		return fmt.Errorf("couldn't create netlink request: %w", err)
+	}
+
+	res, err := c.Execute(req)
+	if err != nil {
+		return fmt.Errorf("couldn't set link down: %w", err)
+	}
+	if len(res) > 1 {
+		return fmt.Errorf("expected 1 message, got %d", len(res))
+	}
+	return nil
+}
+
+func (d *Device) GetBitrate() (uint32, error) {
+	if err := d.updateInfo(); err != nil {
+		return 0, fmt.Errorf("couldn't retrieve bitrate: %w", err)
+	}
+	return d.li.info.BitTiming.Bitrate, nil
+}
+
+func (d *Device) SetBitrate(bitrate uint32) error {
+	c, err := netlink.Dial(unix.NETLINK_ROUTE, &netlink.Config{})
+	if err != nil {
+		return fmt.Errorf("couldn't dial netlink socket: %w", err)
+	}
+	defer c.Close()
+
+	ifi := &ifInfoMsg{}
+	ifi.Index = d.index
```

Comment:
Why set this value on a separate line and not in the struct init on the row above?